### DISCUSSION TITLE
Is There Proper Way to List Defined Model Relationships in Arrays ?

### DISCUSSION THREAD

**Discussion by rupeshstha** [laravel/framework#51754] [Q&A]
I am looking for a proper way to list all defined relationships for a given model in arrays. Currently, I have devised the following method using class reflection to retrieve relationships:

```php
public function getModelRelationships(object $model): array
{
    $relationships = [];
    $modelMethods = (new ReflectionClass($model))->getMethods(ReflectionMethod::IS_PUBLIC);

    foreach ($modelMethods as $method) {
        if (
            $method->class != get_class($model)
            || !empty($method->getParameters())
            || $method->getName() == __FUNCTION__
        ) {
            continue;
        }

        try {
            $return = $method->invoke($model);

            if ($return instanceof Relation) {
                $relationships[] = [
                    $method->getName(),
                    Str::snake(Str::singular($method->getName())),
                ];
            }
        } catch (Exception $exception) {
            // do nothing
        }
    }

    $relationships = array_unique(Arr::flatten($relationships));
    return $relationships;
}
```
This code effectively uses reflection to list relationships directly defined within the model. However, Laravel also allows relationships to be bound through the service container using the `resolveRelationUsing` method, which can be defined in a service provider. This feature is not accounted for in the above method, and thus, it does not capture relationships registered in this manner.

I am seeking guidance on how to properly list all model relationships, including those resolved using the service container. Is there a built-in way in Laravel to achieve this, or could you suggest an approach to enhance the current method to include relationships defined via `resolveRelationUsing`?

**Comment by ghost** [comment]


**Reply by rupeshstha** [reply]
No but i really only need the relationships array not other methods. i know we can do it anyway but i need a proper solution for only fetching list of model relationship.

**Comment by ghost** [comment]


**Reply by rupeshstha** [reply]
Yeah thanks for the reply.  I could do a walk around anyway. I have similar approach where i have a BaseModel and method getBindRelations that returns binded relations with the related model object something like.
```php
use Illuminate\Database\Eloquent\Model;

abstract class BaseModel extends Model
{
    public static function getBindRelations(): array
    {
        return static::$relationResolvers[static::class] ?? [];
    }
}

//  some other class where i want logic
class CacheResolver
{
    public function getModelRelationships(): array
    {
        $relationships = [];
        $modelMethods = (new ReflectionClass($this->model))->getMethods(ReflectionMethod::IS_PUBLIC);
        $bindRelations = $this->model::getBindRelations();

        // Checks relations if model class has relations on it's own class
        foreach ($modelMethods as $method) {
            if (
                $method->class != $this->model::class
                || !empty($method->getParameters())
                || $method->getName() == __FUNCTION__
            ) {
                continue;
            }

            $return = $method->invoke($this->model);
            if ($return instanceof Relation) {
                $relationships[$method->getName()] = [
                    "name" => $method->getName(),
                    "relates_to" => $return->getRelated(),
                ];
            }
        }

        // Checks relations that are resolved from service provider
        foreach ($bindRelations as $method => $bindRelation) {
            $executeClosure = $bindRelation->call($this->model, $this->model);
            if ($executeClosure instanceof Relation) {
                $relationships[$method] = [
                    "name" => $method,
                    "relates_to" => $executeClosure->getRelated(),
                ];
            }
        }

        return $relationships;
    }
} 
```
previous above code was dummy code that could get the picture what i was referring, But i wanted best solution. Maybe we could develop this functionality in next version 😄

**Reply by rupeshstha** [reply]
No mate `relationResolvers` is static property and in my case it's not always good way to add relation name to constant. 🙂

**Comment by staudenmeir** [comment]
This topic is quite complex and there is no universal solution. Here's a related issue: https://github.com/laravel/framework/issues/51592

**Comment by ghost** [comment]


**Reply by rupeshstha** [reply]
Hey mate thanks for the reply but this is not helpful it's off the topic and you can see my above code dose same thing. 🙂 
https://github.com/laravel/framework/discussions/51754#discussioncomment-9729253
